hooks/tk-multi-loader2/tk-3de4_actions.py

```diff
@@ -0,0 +1,289 @@
+
+"""
+Hook that loads defines all the available actions, broken down by publish type.
+
+
+Originally from ``tk-multi-loader2/tk-maya_actions.py``, this file can be
+replaced by:
+
+1. Creating our own copy of ``tk-multi-loader2`` repository
+2. Edit the ``tk-multi-loader2/tk-maya_actions.py`` in there
+3. Point to our ``tk-multi-loader2`` repository in
+   ``env/includes/app_locations.yml:apps.tk-multi-loader2.location``
+"""
+import glob
+from itertools import groupby, izip
+import os
+import re
+import sgtk
+from sgtk.platform.qt import QtCore, QtGui
+
+import tde4
+
+
+HookBaseClass = sgtk.get_hook_baseclass()
+
+
+class FileExistenceError(Exception):
+    """
+    Exception when files don't exist on disk.
+    """
+    def __init__(self, path):
+        """
+        Initialise the class.
+
+        :param str path: The path that doesn't exist.
+        """
+        message = "Could not find file on disk for published file path: {!r}"
+        super(FileExistenceError, self).__init__(message.format(path))
+
+
+def key_func(frames):
+    """
+    Key func for grouping frames by the difference between frames.
+
+    :param tuple frames: The current and next frame number.
+    """
+    return frames[1] - frames[0]
+
+
+def get_frame_numbers(paths):
+    """
+    Get the list of frame numbers from a file sequence.
+
+    :param list(str) paths: The list of paths to extract the frames from.
+
+    :return: The list of frame numbers.
+    """
+    numbers = []
+    frame_pattern = re.compile(r"\.(\d+)\.")
+    for path in paths:
+        match = frame_pattern.search(path)
+        if match:
+            numbers.append(int(match.group(1)))
+    return sorted(numbers)
+
+
+def get_hash_path_and_range_info_from_seq(path):
+    """
+    Get the path sequence in a format that 3DE can read (####), with the start,
+    end and step of the sequence.
+
+    :param str path: The path supplied from shotgun.
+
+    :rtype: tuple(str, int, int, int)
+
+    :raises ValueError: Frame step is not consistent, indicating missing frames.
+    :raises FileExistenceError: The path does not exist on disk.
+    """
+    frame_pattern = re.compile(r"(%0(\d+)d)")
+    frame_match = frame_pattern.search(path)
+    if frame_match:
+        has_frame_spec = True
+        frame_spec = frame_match.group(1)
+        glob_path = path.replace(frame_spec, "*")
+        frame_files = glob.glob(glob_path)
+        numbers = get_frame_numbers(frame_files)
+        steps = []
+        if not frame_files:
+            raise FileExistenceError(path)
+        for step, _ in groupby(izip(numbers[:-1], numbers[1:]), key_func):
+            steps.append(step)
+        if len(steps) > 1:
+            raise ValueError("Inconsistent frame steps")
+        hashed_path = path.replace(frame_spec, "#" * int(frame_match.group(2)))
+        start = min(numbers)
+        end = max(numbers)
+        step = steps[0]
+        return hashed_path, start, end, step
+    return path, 1, 1, 1
```

Comment:
might be better to initialise `start`, `end`, `step` so you only have 1 `return` statement

```suggestion
    start, end, step = 1, 1, 1
    if frame_match:
        has_frame_spec = True
        frame_spec = frame_match.group(1)
        glob_path = path.replace(frame_spec, "*")
        frame_files = glob.glob(glob_path)
        numbers = get_frame_numbers(frame_files)
        steps = []
        if not frame_files:
            raise FileExistenceError(path)
        for step, _ in groupby(izip(numbers[:-1], numbers[1:]), key_func):
            steps.append(step)
        if len(steps) > 1:
            raise ValueError("Inconsistent frame steps")
        path = path.replace(frame_spec, "#" * int(frame_match.group(2)))
        start, end, step = min(numbers), max(numbers), steps[0]
    return path, start, end, step
```